frontend/ui/widget/textboxwidget.lua

```diff
@@ -562,6 +854,25 @@ function TextBoxWidget:getVisibleHeightRatios()
     return low, high
 end
 
+-- Helper function to be used before intanstiating a TextBoxWidget instance
+function TextBoxWidget:getFontSizeToFitHeight(height_px, nb_lines, line_height_em)
+    -- Get a font size that would fit nb_lines in height_px.
+    -- A font with the returned size should then be provided
+    -- to TextBoxWidget:new() (as well as the line_height_em given
+    -- here, as the line_height= property, if not the default).
+    if not nb_lines then
+        nb_lines = 1 -- default to 1 line
+    end
+    if not line_height_em then
+        line_height_em = self.line_height -- (TextBoxWidget default above: 0.3)
+    end
+    -- We do the revert of what's done in :init():
+    --   self.line_height_px = Math.round( (1 + self.line_height) * self.face.size )
+    local font_size = height_px / nb_lines / (1 + line_height_em)
+    font_size = font_size * 1000000 / Screen:scaleBySize(1000000) -- invert scaleBySize
```

Comment:
Doing the proper math in a real matching function in ffi/framebuffer.lua might be worth it?

(i.e., having a real px to pt helper function available).

Reply:
Probably, we used that same trick in coverbrowser some weeks ago.
(But I suck at proper math :)

Reply:
Apparently as simple as basically copying `scaleBySize` but changing the return to `math.floor(px / (size_scale + dpi_scale) * 2)` ;).

Reply:
So, should it be that (scaleBySize() input being pt and not px)?
```diff
-function fb:scaleBySize(px)
+function fb:scaleBySize(pt)
     -- larger screen needs larger scale
@@ -299,3 +299,15 @@ function fb:scaleBySize(px)
     -- scaled positive px should also be positive
-    return math.ceil(px * (size_scale + dpi_scale) / 2)
+    return math.ceil(pt * (size_scale + dpi_scale) / 2)
+end
+
+function fb:unscaleBySize(px)
+    -- Revert of previous function:
+    -- px = scaleBySize(pt)
+    -- pt = unscaleBySize(px)
+    local size_scale = math.min(self:getWidth(), self:getHeight())/600
+    local dpi_scale = size_scale
+    if self.device and self.device.display_dpi ~= self.dpi then
+        dpi_scale = self.dpi / 160
+    end
+    return math.floor(px / (size_scale + dpi_scale) * 2)
 end
```

Reply:
Well, it's not technically points per-se, as that's pretty much set in stone as `px = round(dpi / 72.0 * pt)`, which isn't what we're doing.

It's more something like absolute pixels or whatever (I only mentioned point earlier because the theory is similar ;)).

Reply:
OK, so keeping using px in scaleBySize (pt means portuguese to me anyway:).

Reply:
@Frenzie: please confirm you're ok with that, before I add it to my base PR and trigger the whole CI:
```lua
function fb:unscaleBySize(px)
    -- Invert of fb:scaleBySize(px)
    local size_scale = math.min(self:getWidth(), self:getHeight())/600
    local dpi_scale = size_scale
    if self.device and self.device.display_dpi ~= self.dpi then
        dpi_scale = self.dpi / 160
    end
    return math.floor(px / (size_scale + dpi_scale) * 2)
end
```

Reply:
Link to the originals for convenience:

https://github.com/koreader/koreader-base/blob/c140b752e16163ee8836214787caf7343bc4a579/ffi/framebuffer.lua#L285-L301

It looks sane enough at a glance, so I'm okay with it provided you checked it actually outputs the right values. ;-)

Reply:
Actually you can do that by simply adding an extra line here:

https://github.com/koreader/koreader-base/blob/c140b752e16163ee8836214787caf7343bc4a579/spec/unit/framebuffer_spec.lua#L23-L29

Reply:
Well... given the rounding we do, unit tests fails...
```
        fb:setDPI(167)
        assert.are.equals(20, fb:scaleBySize(20 * 1000000 / fb:scaleBySize(1000000)))
        assert.are.equals(20, fb:scaleBySize(fb:unscaleBySize(20)))
        fb:setDPI(167 * 3)
        assert.are.equals(18, fb:scaleBySize(18 * 1000000 / fb:scaleBySize(1000000)))
        assert.are.equals(18, fb:scaleBySize(fb:unscaleBySize(18)))
```
whether I use floor or ceil or round, one of these fail. The non rounding 1000000 stuff doesn't fail.
So:
```
    -- Because of possible rounding error, as we only pass and
    -- return integers, scaleBySize(unscaleBySize(N)) != N
    -- In the context this is used, we prefer getting a smaller
    -- one. So use math.floor()
    return math.floor(px / (size_scale + dpi_scale) * 2)
```
?
Or return a float, and let callers do the rounding they want?
And how to unit-test that if we can't check equality? `> 0.9*N and < N` ?

(And please, stop bringing me into your bidi math games! :)

Reply:
Well, no frankly, not my war.
That api would look suspicious, and the unit tests even more fuzzy - so better to not have it, and let callers do they fuzzy math themselves.

Reply:
You didn't mention how they fail precisely. You could just do an imprecise test, like:

```lua
local unscaled_size = fb:scaleBySize(fb:unscaleBySize(18))
assert.is_true(unscaled_size > 17 && unscaled_size < 19)
```

Or given that we control the test, something like `assert.are.equals(19, fb:scaleBySize(fb:unscaleBySize(18)))` would also work. The point of the unit test is primarily to ensure we still get the result we expect in the future.

Reply:
Yeah, 19 vs 20, stuff like that. And may be more than +/-1 with a high dpi value.
But this function is ridiculous if it's unprecise. And if the caller wants it rounded up, it would need to be some twins unscaleBySizeRoundedUp/unscaleBySizeRoundedDown.
I get more precise results with the 1000000 stuff. May be it's not needed by the 2 callers (textwidget/menu, & coverbrowser), but I don't want to investigate if they would be ok with floor or a little lower, or if they would prefer ceil and a little higher- as it's all just fine and tested with precise 1000000 stuff.

Reply:
Oh, yeah, round trips are going to be hard to keep stable without preserving the precise value (i.e., the float).

I'm not quite sure by which arcane trickery this currently appears to be the case with the integer trickery.

Reply:
> I'm not quite sure by which arcane tricke

Because I lied a bit and did not do any rounding in `18 * 1000000 / fb:scaleBySize(1000000)` may be :) But still.

Reply:
That said, I'm slightly miffed by the fact that it appears to behave in Python...

```py
#!/usr/bin/env python3

import math

dpi = 167
width = 600
height = 800

def scaleBySize(px):
   # larger screen needs larger scale
   size_scale = min(width, height)/600.0
   # if users custom screen dpi, also scale by dpi
   dpi_scale = size_scale

   dpi_scale = dpi / 160.0

   # scaled positive px should also be positive
   scaled_px = px * (size_scale + dpi_scale) / 2.0
   return (scaled_px, math.ceil(scaled_px))

def unScaleBySize(px):
   # larger screen needs larger scale
   size_scale = min(width, height)/600.0
   # if users custom screen dpi, also scale by dpi
   dpi_scale = size_scale

   dpi_scale = dpi / 160.0

   # scaled positive px should also be positive
   unscaled_px = px / (size_scale + dpi_scale) * 2.0
   return (unscaled_px, math.floor(unscaled_px))

print("scaleBySize(500): {}".format(scaleBySize(500.0)))
print("unScaleBySize(511): {}".format(unScaleBySize(511.0)))

print("scaleBySize(20): {}".format(scaleBySize(20.0)))
print("unScaleBySize(21): {}".format(unScaleBySize(21.0)))

dpi = 167 * 3

print("scaleBySize(18): {}".format(scaleBySize(18.0)))
print("unScaleBySize(38): {}".format(unScaleBySize(38.0)))

print("18 via * 10**6: {}".format(scaleBySize(18 * 1000000.0 / scaleBySize(1000000.0)[1])))
```

```
scaleBySize(500): (510.93750000000006, 511)
unScaleBySize(511): (500.06116207951067, 500)
scaleBySize(20): (20.4375, 21)
unScaleBySize(21): (20.550458715596328, 20)
scaleBySize(18): (37.18125, 38)
unScaleBySize(38): (18.3963691376702, 18)
18 via * 10**6: (17.999999999999996, 18)
```

The emulator defaults to 600x800 @ 167 dpi, right?

Reply:
> 
> The emulator (and as such the testsuite) defaults to 600x800 @ 167 dpi, right?

Possibly not, otherwise scaleBySize would do nothing (as min(600, 800)/600 == 1), and if dpi == device_dpi that's the scale factor).

Reply:
Looks like this test does not  have any screen h/w, so possibly only using dpi_scale.

Or we can just have that unscaleBySize return a float, and document that the caller is responsible to do a floor or ceil depending on what he prefers.
That way, the unit tests would be always fine: `assert.are.equals(18, fb:scaleBySize(fb:unscaleBySize(18)))`

Reply:
Not a Python quirk ;).

```lua
#!/usr/bin/env luajit

dpi = 167
width = 600
height = 800

function scaleBySize(px)
   -- larger screen needs larger scale
   size_scale = math.min(width, height)/600
   -- if users custom screen dpi, also scale by dpi
   dpi_scale = size_scale

   dpi_scale = dpi / 160

   -- scaled positive px should also be positive
   scaled_px = px * (size_scale + dpi_scale) / 2
   return {scaled_px, math.ceil(scaled_px)}
end

function unScaleBySize(px)
   -- larger screen needs larger scale
   size_scale = math.min(width, height)/600
   -- if users custom screen dpi, also scale by dpi
   dpi_scale = size_scale

   dpi_scale = dpi / 160

   -- scaled positive px should also be positive
   unscaled_px = px / (size_scale + dpi_scale) * 2
   return {unscaled_px, math.floor(unscaled_px)}
end

---

function tprint(tbl, indent)
  if not indent then indent = 0 end
  for k, v in pairs(tbl) do
    formatting = string.rep("  ", indent) .. k .. ": "
    if type(v) == "table" then
      print(formatting)
      tprint(v, indent+1)
    elseif type(v) == 'boolean' then
      print(formatting .. tostring(v))
    else
      print(formatting .. v)
    end
  end
end

print("scaleBySize(500)")
tprint(scaleBySize(500))
print("unScaleBySize(511)")
tprint(unScaleBySize(511))

print("scaleBySize(20)")
tprint(scaleBySize(20))
print("unScaleBySize(21)")
tprint(unScaleBySize(21))

dpi = 167 * 3

print("scaleBySize(18)")
tprint(scaleBySize(18))
print("unScaleBySize(38)")
tprint(unScaleBySize(38))

print("18 via * 10**6")
tprint(scaleBySize(18 * 1000000 / scaleBySize(1000000)[2]))
```

```
scaleBySize(500)
1: 510.9375
2: 511
unScaleBySize(511)
1: 500.06116207951
2: 500
scaleBySize(20)
1: 20.4375
2: 21
unScaleBySize(21)
1: 20.550458715596
2: 20
scaleBySize(18)
1: 37.18125
2: 38
unScaleBySize(38)
1: 18.39636913767
2: 18
18 via * 10**6
1: 18
2: 18
```

Reply:
TL;DR: Smells like yet another case of "I don't know what the hell's happening inside the testsuite" ;).

Reply:
(GH is still being clunky, I've only just now got your previous answers interleaved in the thread -_-").

Reply:
GH has recently been "helpfully" hiding comments behind a "More comments" type thing for me. D:

The presumed default *used* to be 800x600 @ 167 but that was changed to 160 for easier math. See https://github.com/koreader/koreader-base/commit/c978857658fd5343f2fe90534c0d998cdead9aa8

Otherwise the unit test would make no sense! ;-)

```lua
        fb:setDPI(167)
        assert.are.equals(31, fb:scaleBySize(30))

        fb:setDPI(167 * 3)
        assert.are.equals(62, fb:scaleBySize(30))
```

Reply:
Or at least, I supported it for that reason: https://github.com/koreader/koreader-base/pull/843#issuecomment-469597361